_posts/2018-11-29-async-update.md

```diff
@@ -0,0 +1,201 @@
+---
+layout: post
+title: "Async in Rust, circa 2018"
+date: 2018-11-29
+author: Aaron Turon
+---
+
+The Rust 2018 Edition, and the end of 2018, are fast approaching! While we
+didn't manage to ship async/await as part of the edition itself, the
+community has made quite a lot of progress toward that goal. This post
+summarizes the state of play, and announces the publication several crates
+intended to facilitate use of async/await on the nightly ecosystem.
+
+# Why async/await
+
+Before delving into the current status, it's worth taking a moment to recap
+the core motivations for async/await, and its special importance for Rust.
+
+Async/await notation is a way of making asynchronous programming more
+closely resemble synchronous programming. To see how this works, consider
+`Read::read` in `std::io`:
+
+```rust
+fn read(&mut self, buf: &mut [u8]) -> Result<usize, std::io::Error>
+```
+
+This synchronous method *blocks the current thread* until data has been read
+into `buf`, then says how many bytes were read. We can build on this method
+to implement `read_exact`, a method that *continues* reading until the buffer
+is filled:
+
+```rust
+fn read_exact<T: Read>(input: &mut T, buf: &mut [u8]) -> Result<(), std::io::Error> {
+    let mut cursor = 0;
+    while cursor < buf.len() {
+        cursor += input.read(&mut buf[cursor..])?;
+    }
+}
+```
+
+In the asynchronous world, we want to perform similar operations, but rather
+than blocking the current thread we want to leave it free to do other work
+while the I/O operations complete asychronously. But actually programming
+*directly* in that way is incredibly difficult. What we want is to program
+*as if* I/O operations will block the current thread, but have supporting
+infrastructure transform this code into more efficient asynchronous execution.
```

Comment:
"but have the compiler transform the code into an efficient asynchronous implementation"?